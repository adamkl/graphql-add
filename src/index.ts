const debug = require("debug")("graphql-add");
import { map, filter, uniqWith, flow } from "lodash/fp";

import {
  DocumentNode,
  parse,
  print,
  visit,
  FieldNode,
  SelectionSetNode,
  SelectionNode,
  FragmentSpreadNode,
  FragmentDefinitionNode
} from "graphql";
import { inspect } from "util";

const logObj = (document: any) => {
  debug(inspect(document, { depth: 20, colors: true }));
};

type NamedSelectionNode = FieldNode | FragmentSpreadNode;

const isNamedNode = (node: SelectionNode): boolean => {
  return (node as NamedSelectionNode).name !== undefined;
};

export const combineQueries = (queries: (string | DocumentNode)[]) => {
  debug("combining queries");
  if (queries.length === 0) {
    return "";
  }

  const documents = queries.map(q => (typeof q === "string" ? parse(q) : q));

  if (documents.length < 2) {
    return print(documents[0]);
  }
  documents.forEach(d => debug(inspect(d, { depth: 20, colors: true })));
  let merged = documents.shift();
  while (documents.length > 0) {
    let incoming = documents.shift();
    merged = visit(merged, {
      Document: node => {
        let incomingFragments: FragmentDefinitionNode[] = [];
        visit(incoming, {
          FragmentDefinition: incomingNode =>
            incomingFragments.push(incomingNode)
        });
        return {
          ...node,
          definitions: [...node.definitions, ...incomingFragments]
        };
      },
      SelectionSet: {
        leave: (mNode, mKey, mParent, mPath, mAncestors) => {
          const mergedParent = mParent as FieldNode;
          let found: SelectionSetNode;
          incoming = visit(incoming, {
            SelectionSet: {
              leave: (iNode, iKey, iParent, iPath, iAncestors) => {
                const incomingParent = iParent as FieldNode;
                debug({
                  mergedParent,
                  incomingParent
                });
                if (mergedParent.name.value === incomingParent.name.value) {
                  found = iNode;
                  return null;
                }
              }
            },
            Field: {
              leave: iNode => {
                if (iNode.selectionSet === null) {
                  return null;
                }
              }
            }
          });
          if (!found) {
            return;
          }
          const selections = flow(
            filter(isNamedNode),
            map(n => n as NamedSelectionNode),
            uniqWith((a, b) => a.name.value === b.name.value)
          )([...mNode.selections, ...found.selections]);

          const combined = {
            kind: "SelectionSet",
            selections
          };
          //logObj({ mNode, found, combined });
          return combined;
        }
      }
    });
  }
  merged = visit(merged, {
    OperationDefinition: node => {
      return {
        ...node,
        name: {
          kind: "Name",
          value: "mergedQuery"
        }
      };
    }
  });

  return print(merged);
};
