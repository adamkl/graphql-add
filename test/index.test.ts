import { combineQueries } from "../src/index";
import { print, parse } from "graphql";

test("it should combine queries", () => {
  const queries = [
    `
query q1 {
  person {
    name
  }
}
  `,
    `
query q2 {
  person {
    age
  }
}
  `
  ];

  const rv = combineQueries(queries);
  expect(rv).toMatchSnapshot();
});

test("it should combine more than 2 queries", () => {
  const queries = [
    `
query q1 {
  person {
    name
  }
}
  `,
    `
query q2 {
  person {
    age
  }
}
  `,
    `
query q3 {
  person {
    dob
  }
}
  `
  ];

  const rv = combineQueries(queries);
  expect(rv).toMatchSnapshot();
});

test("it should remove duplicate fields", () => {
  const queries = [
    `
query q1 {
  person {
    name
  }
}
  `,
    `
query q2 {
  person {
    name
  }
}
  `,
    `
query q3 {
  person {
    age
  }
}`
  ];

  const rv = combineQueries(queries);
  expect(rv).toMatchSnapshot();
});

test("it should combine nested queries", () => {
  const queries = [
    `
query q1 {
  person {
    address {
      street
    }
  }
}
  `,
    `
query q2 {
  person {
    address {
      city
    }
  }
}
  `
  ];

  const rv = combineQueries(queries);
  expect(rv).toMatchSnapshot();
});

test("it should combine nested queries with fields at different levels", () => {
  const queries = [
    `
query q1 {
  person {
    firstName
    address {
      street
    }
  }
}
  `,
    `
query q2 {
  person {
    lastName
    address {
      city
    }
  }
}
  `
  ];

  const rv = combineQueries(queries);
  expect(rv).toMatchSnapshot();
});

test("it should combine queries with fragment spread", () => {
  const queries = [
    `
query q1 {
  person {
    firstName
    address {
      street
    }
  }
}
  `,
    `
query q2 {
  person {
    ...personFragment
    address {
      city
    }
  }
}

fragment personFragment on Person {
  lastName
}
  `
  ];

  const rv = combineQueries(queries);
  expect(rv).toMatchSnapshot();
});

test("it should combine queries with multiple fragment spreads", () => {
  const queries = [
    `
query q1 {
  person {
    firstName
    address {
      street
    }
  }
}
  `,
    `
query q2 {
  person {
    ...personFragment
    address {
      city
      ...cityFragment
    }
  }
}

fragment personFragment on Person {
  lastName
}

fragment cityFragment on City {
  country
}
  `,
    `
query q3 {
  person {
    ...dobFragment
  }
}

fragment dobFragment on Person {
  dob
}
  `
  ];

  const rv = combineQueries(queries);
  expect(rv).toMatchSnapshot();
});
