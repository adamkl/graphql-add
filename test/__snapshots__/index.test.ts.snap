// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should combine more than 2 queries 1`] = `
"query mergedQuery {
  person {
    name
    age
    dob
  }
}
"
`;

exports[`it should combine nested queries 1`] = `
"query mergedQuery {
  person {
    address {
      street
      city
    }
  }
}
"
`;

exports[`it should combine nested queries with fields at different levels 1`] = `
"query mergedQuery {
  person {
    firstName
    address {
      street
      city
    }
    lastName
  }
}
"
`;

exports[`it should combine queries 1`] = `
"query mergedQuery {
  person {
    name
    age
  }
}
"
`;

exports[`it should combine queries with fragment spread 1`] = `
"query mergedQuery {
  person {
    firstName
    address {
      street
      city
    }
    ...personFragment
  }
}

fragment personFragment on Person {
  lastName
}
"
`;

exports[`it should combine queries with multiple fragment spreads 1`] = `
"query mergedQuery {
  person {
    firstName
    address {
      street
      city
      ...cityFragment
    }
    ...personFragment
    ...dobFragment
  }
}

fragment personFragment on Person {
  lastName
}

fragment cityFragment on City {
  country
}

fragment dobFragment on Person {
  dob
}
"
`;

exports[`it should remove duplicate fields 1`] = `
"query mergedQuery {
  person {
    name
    age
  }
}
"
`;
